[HLSL][RootSignature] Implement Parsing of Descriptor Tables

inbelic opened the pull request:
Implementation of parsing the root signature in its token form into the in-memory representation of `RootElement`s. Specifically, it handles all Descriptor Table related tokens and parses these into the `DescriptorTable` and `DescriptorTableClause` struct representation.

This will define all the helper methods of Root Signature Parser needed for parsing of general elements.

Note that this parser definition will be completely self-contained and can be invoked externally when creating the `HLSLRootSignatureAttr`

- Defines the in-memory representation of `RootElement`, `DescriptorTable` and `DescriptorTableClause` in `HLSLRootSignatures.h`
- Defines `RootSignatureParser` and provide a public `Parse` that will take an input `RootSignatureLexer` and generate a list of `RootElement`s
- Implements the internal methods and helper methods required to parse the different types of `RootElements`
- Adds unit testing in `ParseHLSLRootSignatureTest.cpp`

Resolves #126569

inbelic commented:
Please ignore the first commit. It is a duplicate of https://github.com/llvm/llvm-project/pull/122981

inbelic commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -89,6 +91,72 @@ class RootSignatureLexer {
+  bool Parse();
+
+private:
+  bool ReportError(); // TODO: Implement this to report error through Diags
```
I have left this as a todo just to help with pr size, as it will require adding a good chunk of additional testing

llvm-beanz replied:
I don't think this should be separate. It's going to be really hard to ensure that any follow-up that adds error reporting properly covers all the cases.

V-FEXrt commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -148,6 +148,333 @@ bool RootSignatureLexer::LexToken(RootSignatureToken &Result) {
+bool RootSignatureParser::ReportError() { return true; }
+
+bool RootSignatureParser::Parse() {
+  CurTok--; // Decrement once here so we can use the ...ExpectedToken api
```
This is UB right? You are decrementing the `CurTok` out of its actual bounds

inbelic replied:
Correct. I didn't think that it was undefined behaviour if you guarantee not to dereference while out of bounds, but that is incorrect. And bad practice either way, so updated the loop to handle that the first iteration will have the pointer be 1 ahead of expected.

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -148,6 +148,333 @@ bool RootSignatureLexer::LexToken(RootSignatureToken &Result) {
+
+  // Iterate as many DescriptorTableClaues as possible
+  bool HasComma = true;
+  while (!TryConsumeExpectedToken({TokenKind::kw_CBV, TokenKind::kw_SRV,
```
Good catch. Yep, it would otherwise allow invalid signatures of the form: `CBV() UAV()` where there is no comma specified

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -148,6 +148,333 @@ bool RootSignatureLexer::LexToken(RootSignatureToken &Result) {
+  // Iterate as many RootElements as possible
+  bool HasComma = true;
+  while (HasComma &&
+         !TryConsumeExpectedToken(ArrayRef{TokenKind::kw_DescriptorTable})) {
```
Updated to have an empty edge case and added test for it

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -148,6 +148,333 @@ bool RootSignatureLexer::LexToken(RootSignatureToken &Result) {
+  if (ParseRegister(Clause.Register))
+    return true;
+
+  // Start parsing the optional parameters
```
My understanding of the grammar specified [here](https://github.com/llvm/wg-hlsl/blob/main/proposals/0002-root-signature-in-clang.md#root-signature-grammar) was that they need to be in the given order. Although looking at DXC's implementation it seems they can be specified in any order.

inbelic replied:
Confirmed that we should accept the parameters in any order. So will need to refactor this.

llvm-beanz reviewed (commented):
I think that the way you're breaking up this change is sub-optimal from a review perspective. You've added a lot of code that partially handles parsing a very complex root signature. The problem is that to complete this implementation you're going to go back over this code over and over again fleshing it out, and from a reviewer's perspective we're going to need to keep paging back in extra context.

If instead you started with a much simpler root signature (even just an empty one), but implement more complete handling for it, we can review that and incrementally build up without revisiting the same code over and over again in each subsequent patch.

inbelic commented:
Sounds good, and I appreciate the feedback. I will restructure the changes to be of smaller granularity, which will be better self-contained and directly include their diagnostics testing.

damyanp commented on llvm/include/llvm/Frontend/HLSL/HLSLRootSignature.h:
```diff
@@ -0,0 +1,140 @@
+};
+
+// Models RootElement : DescriptorTable | DescriptorTableClause
+struct RootElement {
```
Is there a reason not to use `std::variant`?

inbelic replied:
Nope, good idea! Switched

damyanp commented on llvm/include/llvm/Frontend/HLSL/HLSLRootSignature.h:
```diff
@@ -0,0 +1,140 @@
+enum class ClauseType { CBV, SRV, UAV, Sampler };
+struct DescriptorTableClause {
+  ClauseType Type;
+  Register Register;
```
It looks like there's nothing enforcing Register's initialization?  Since this struct has a constructor, I'd expect to get a fully initialized object back when I construct it.

inbelic replied:
Switched to use a non-constructor function to avoid implying that the register would be initialized.

inbelic committed:
add support for optional parameters

- use `numDescriptors` as example

inbelic committed:
add space optional parameter

- demonstrate can specify in any order

inbelic committed:
define custom paramtype for param parsing - descriptor range offset

inbelic committed:
add support for parsing enums

- uses Shader Visibility as an example for demonstrated test cases

inbelic committed:
add support for parsing flags

- use descriptor range flags as an example

inbelic commented:
Updated to include diagnostics output and the relevant testing.

I have rebased to split the pr into smaller incremental changes, to help with review. And I have taken care to address all previous comments. Unfortunately the rebase causes these comments to get lost :/

I have implemented the suggestion to use `std::variant` to model the `RootElement` data-type, re-used the `dxil::ResourceClass` for `ClauseType`.

Notably though, there was some bigger changes introduced to handle that the optional parameters can be specified in any order.

joaosaffran commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -169,5 +220,399 @@ bool RootSignatureLexer::LexToken(RootSignatureToken &Result) {
+    return true;
+
+  // Dispatch onto the correct parse method
+  switch (CurTok->Kind) {
```
I was expecting this switch to happen earlier in this method. What is the reason to ensure that descriptor tables are the first element of Root Signatures? For example, this is valid `SRV(t0), DescriptorTable(CBV(b0, numDescriptors=5))` and doesn't start with a `DescriptorTable`.

inbelic replied:
This pr only adds support for `DescriptorTable` as a root element. So there is no added support for a root descriptor range at the moment. A future pr that will handle the `RootCBV, etc` would be added to this switch statement and then dispatched onto its own method.

joaosaffran commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -169,5 +220,399 @@ bool RootSignatureLexer::LexToken(RootSignatureToken &Result) {
+}
+
+bool RootSignatureParser::ParseDescriptorTableClause() {
+  if (ConsumeExpectedToken({TokenKind::kw_CBV, TokenKind::kw_SRV,
+                            TokenKind::kw_UAV, TokenKind::kw_Sampler}))
+    return true;
+
+  DescriptorTableClause Clause;
+  switch (CurTok->Kind) {
+  case TokenKind::kw_CBV:
+    Clause.Type = ClauseType::CBuffer;
+    break;
+  case TokenKind::kw_SRV:
+    Clause.Type = ClauseType::SRV;
+    break;
+  case TokenKind::kw_UAV:
+    Clause.Type = ClauseType::UAV;
+    break;
+  case TokenKind::kw_Sampler:
+    Clause.Type = ClauseType::Sampler;
+    break;
+  default:
+    llvm_unreachable("Switch for an expected token was not provided");
+    return true;
+  }
+  Clause.SetDefaultFlags();
```
It might be interesting to move this to its own method, since it can be reused when handling root descriptors.

joaosaffran reviewed (commented):
How are we planning to handle the parsing of different root signatures versions ? Not sure if I miss it

bogner commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -80,6 +85,99 @@ class RootSignatureLexer {
+  bool ParseParam(rs::ParamType Ref);
+
+  // Parse as many optional parameters as possible in any order
+  bool
+  ParseOptionalParams(llvm::SmallDenseMap<TokenKind, rs::ParamType> RefMap);
```
Do you really want to be passing a `SmallDenseMap` by value here? This and the other similar APIs should probably be using references.

bogner commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -80,6 +85,99 @@ class RootSignatureLexer {
+                      const SmallVector<RootSignatureToken> &Tokens,
+                      DiagnosticsEngine &Diags);
+
+  // Iterates over the provided tokens and constructs the in-memory
+  // representations of the RootElements.
+  //
+  // The return value denotes if there was a failure and the method will
+  // return on the first encountered failure, or, return false if it
+  // can sucessfully reach the end of the tokens.
```
If you use `///` instead of `//` for the doc comments on the methods of this class they'll show up in the [LLVM doxygen](https://llvm.org/doxygen/).

bogner commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -148,6 +148,347 @@ bool RootSignatureLexer::LexToken(RootSignatureToken &Result) {
+    return ParseDescriptorTable();
+  default:
+    llvm_unreachable("Switch for an expected token was not provided");
+    return true;
```
`llvm_unreachable` doesn't return (it is annotated as a "noreturn" function), so the `return true` is unreachable code here. This won't generally lead to the parser erroring out in practice, because the program will crash or hit UB due to the `llvm_unreachable`. Given that we expect to have actual error handling for the token kind being appropriate just above, this is probably okay, but it does mean that the `return true` is unnecessary.

inbelic committed:
[HLSL][RootSignature] Parse empty Descriptor Table Root Element

- defines the Parser class and an initial set of helper methods to
support consuming tokens. functionality is demonstrated through a simple
empty descriptor table test case
- defines an initial in-memory representation of a DescriptorTable
- implements a test harness that will be used to validate the correct
diagnostics are generated. it will construct a dummy pre-processor with
diagnostics consumer to do so

inbelic committed:
add empty Descriptor Table Clause

inbelic committed:
add Register and implicitly unsigned integers

inbelic committed:
clang formatting

inbelic committed:
review: remove unreachable return true

inbelic committed:
review: clean-up casting of test values

inbelic committed:
self-review: update doc

inbelic committed:
self-review: move token spellings to definition

inbelic commented on clang/unittests/Parse/ParseHLSLRootSignatureTest.cpp:
```diff
@@ -306,4 +307,254 @@ TEST_F(ParseHLSLRootSignatureTest, InvalidLexRegNumberTest) {
+  Consumer->SetNoDiag();
+
+  ASSERT_FALSE(Parser.Parse());
+  ASSERT_EQ((int)Elements.size(), 0);
```
Updated this and other casting to use the `u` notation.

inbelic replied:
Updated this and other casting to use the `u` notation on the literals

inbelic commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -93,6 +96,108 @@ class RootSignatureLexer {
+  bool ParseOptionalParams(
+      llvm::SmallDenseMap<TokenKind, llvm::hlsl::rootsig::ParamType> &RefMap);
+
+  // Common parsing helpers
```
I restructured the comments a bit so I don't think this applies anymore.

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -162,5 +219,371 @@ std::optional<RootSignatureToken> RootSignatureLexer::PeekNextToken() {
+    : Elements(Elements), Lexer(Lexer), Diags(Diags) {}
+
+bool RootSignatureParser::Parse() {
+  // Handle edge-case of empty RootSignature()
```
My understanding is that from the perspective of parsing, an empty root signature would be valid. Instead an error is raised later when validating the structure. Presumably because not all the resources have a binding or so

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -162,5 +219,371 @@ std::optional<RootSignatureToken> RootSignatureLexer::PeekNextToken() {
+  case TokenKind::kw_DescriptorTable:
+    return ParseDescriptorTable();
+  default:
+    llvm_unreachable("Switch for an expected token was not provided");
```
At the moment, we follow the pattern of `ConsumeExpectedToken({TokenKind1, ..., TokenKindN})` and then switch on the token kinds that were in that statement. `ConsumeExpectedToken` guards against any other token kind. So any new token kind is added we will never reach the default here, unless it is also added to the `ConsumeExpectedToken` statement.

Maybe there is a way we could define the macros to generate sub-sets of tokens? Although, I am of the opinion the current method guards against this sufficiently.

inbelic commented:
Rebasing onto the lexer changes that have been added to main. Since there were some disruptive review changes in the lexer (removing parsing of integers, removing any error handling and moving the lexer to the Lex subdir) that we are rebasing on, it was quite difficult to apply those changes in a nice way here without have a rebase.

So apologies for having to rebase here. What has actually changed in the rebase, is the addition of the `HandleUIntLiteral` function and a change in how `ConsumeExpectedToken`/`TryConsumeExpectedToken` work as they no longer need to propagate lexer errors. Other than that, it is just documentation updates from the previous review and some small fixups.

inbelic commented:
Also I am also unsure why it triggered a review from all the code owners. Is it because I force-pushed before changing the target branch?

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,417 @@
+    Clause.Type = ClauseType::Sampler;
+    break;
+  default:
+    llvm_unreachable("Switch for an expected token was not provided");
```
Forwarding this for visibility: https://github.com/llvm/llvm-project/pull/122982#discussion_r1953996279

bob80905 commented on clang/unittests/Parse/ParseHLSLRootSignatureTest.cpp:
```diff
@@ -0,0 +1,375 @@
+//=== ParseHLSLRootSignatureTest.cpp - Parse Root Signature tests ---------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
```
clang/lib/Parse/ParseHLSLRootSignature.cpp lacks a License statement at the top of its file.
I'm not sure whether every new cpp file needs a declaration of a license? I bet @beanz knows.

llvm-beanz replied:
Every new file should get the license header.

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,417 @@
+    return false;
+
+  // Report unexpected token kind error
+  Diags().Report(CurToken.TokLoc, diag::err_hlsl_rootsig_unexpected_token_kind)
```
The diagnostic approach here seems quite complicated and inflexible. It doesn't allow for a context-dependent diagnostic.

Clang's approach in the Parser is to have a `TryConsumeToken` method, that returns success/failure, and the caller is then responsible for providing the appropriate diagnostic which allows much better control over the diagnostic text.

inbelic replied:
I think this does effectively using a similar pattern (albeit with different names, which is confusing) as Clang, this function would be equivalent to the `ExpectAndConsume` function in the Clang's implementation.

`ExpectAndConsume` does however allow to pass down custom context specific diag messages. I have added a quick prototype commit to allow using custom diag messages to `ConsumeExpectedToken`.

Is this what you had intended? Do you think we should also change the function names to align with the Clang parser?

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,417 @@
+#include "clang/Parse/ParseHLSLRootSignature.h"
```
This is the file with the missing license header that @bob80905 pointed out.

inbelic committed:
review: add liscence

inbelic committed:
review prototype: allow for more context to ConsumeNextToken error

inbelic committed:
remove use of strings as context and use a previous context token instead

- updates the error messages for missing '(' and ')' tokens with this
approach

inbelic committed:
add diag context when incorrectly specifiying a parameter

inbelic committed:
add diag context when incorrectly given value to parameter

inbelic committed:
add end of stream diag

inbelic committed:
clang-formatting

inbelic commented:
This PR needs to be reworked with recent updates to the root signature specs. Mainly to orchestrate specifying mandatory and optional parameters in any order.

Since the pr has become too large for reviewability, I will open a collection of further prs that implement these changes with the rework incorporated.

The first of which is here: https://github.com/llvm/llvm-project/pull/133302
